File: Leetcode/two-sum/twoSum--good.ts
/**
 * https://leetcode.com/problems/two-sum/
 * Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
 * You may assume that each input would have exactly one solution, and you may not use the same element twice.
 * You can return the answer in any order.
 *
 * Personal note: this one should only be O(n)
 *
 * Runtime: 80ms (faster than 96%)
 * Memory: 39.2mo (less than 99%)
 */
function twoSum(nums: number[], target: number): number[] {
  const numsMap: any = {};
  for(let indexOfTestedNumber = 0; indexOfTestedNumber < nums.length; indexOfTestedNumber++) {
    const currentNumber = nums[indexOfTestedNumber];
    const quantityLeft = target - currentNumber;
    if(typeof numsMap[currentNumber] !== "undefined") return ([indexOfTestedNumber, numsMap[currentNumber]]);
    numsMap[quantityLeft] = indexOfTestedNumber;
  }
  return [];
}

File: Leetcode/two-sum/twoSum--good.py
# https://leetcode.com/problems/two-sum/ Given an array of integers nums and an integer target, return indices of the
# two numbers such that they add up to target. You may assume that each input would have exactly one solution,
# and you may not use the same element twice. You can return the answer in any order.
#
# Runtime: 44ms (faster than 82%)
# Memory: 14mb (less than 72%)

class Solution:
    def twoSum(self, nums, target):
        remains_map = {}
        for i, num in enumerate(nums):
            if num in remains_map:
                return [i, remains_map[num]]
            remains_map.update({target - num: i})
